Docs for Laravel style relations (#176)
wintercms/docs: database/relations.md

Database tables are often related to one another. For example, a blog post may have many comments, or an order could be related to the user who placed it. Winter makes managing and working with these relationships easy and supports several different types of relationships.

> > **NOTE:** If you are selecting specific columns in your query and want to load relationships as well, you need to make sure that the columns that contain the keying data (i.e. `id`, `foreign_key`, etc) are included in your select statement. Otherwise, Winter cannot connect the relations.

Reviewer: Is this no longer true?

Author: Nope, fairly certain that Laravel connects the required foreign keys regardless of any `select()` calls limiting the columns. Probably worth a unit test for this though :)

Reviewer: @bennothommo this is the original issue that prompted that note to be included, if you could do up a unit test that would be fantastic ❤️: https://github.com/octobercms/october/issues/2944

Author: @LukeTowers my mistake - Laravel doesn't do this:

![image](https://github.com/user-attachments/assets/dbb7513f-ba5c-4797-9fa1-85bf3f7e8c4c)

It does when retrieving the relations directly, but if you're creating a query and you don't include the necessary ID fields, it will prevent the relations from being linked. I will restore the note, but I'll move it down to the eager loading section as Laravel have done.

Winter provides two methods of defining model relationships. Both provide the same level of functionality - which one you use is entirely up to your own preferences.

> - Property-based array configuration in the model class, available in all versions of Winter CMS (Property style)
> - Relation methods defined in the model class, synonymous with Laravel, available since Winter CMS v1.2.5 (Method style)

Reviewer: ```suggestion
- Property-based array configuration in the model class (Property style)
- Relation methods defined in the model class, [synonymous with Laravel](https://laravel.com/docs/9.x/eloquent-relationships#defining-relationships) (Method style)
```

### Property style relationship definition

Model relationships can be defined as properties on your model classes. An example of defining relationships:

```php
class User extends Model
{
    public $hasOne = [

>         'profile' => 'Acme\Blog\Models\Profile',

Reviewer: ```suggestion
        'profile' => \Acme\Blog\Models\Profile::class,
```

    ];

    public $hasMany = [

>         'posts' => 'Acme\Blog\Models\Post',

Reviewer: ```suggestion
        'posts' => \Acme\Blog\Models\Post::class,
```

Author: I'd like to keep one example of a string like this, to show it's accepted, even if it's not our "preferred" way.

Reviewer: Meh, I'd prefer not to. If someone wants to do it that way and it works for them, good for them. The documentation should be showing people how we expect things to be done; and a string literal working as a class reference is a language feature, so I think it's unnecessary that we provide an example of it in our docs just for completeness sake. Technically `array('item1', 'item2')` is also a supported way of doing things, but there's no way I want to be encouraging that behaviour.

The ::class constant is extremely helpful, both in making it obvious when you have a typo and making code navigation in editors / IDEs much easier / faster.

];
```

### Method style relation definition

> > **Note:** Method style relation definition is available from Winter v1.2.5.

Reviewer: ```suggestion
> **NOTE:** Method style relation definition is available from Winter v1.2.7.
```
Specific version to be updated once we merge the PR

}
```

The following chained methods are available to define additional parameters about the relation:

Method | Description
------ | -----------
`->dependent()` | Makes this relation "dependent" on the primary model. The related model records will be deleted when the primary model is deleted. This is only available for the following relation types: `attachOne`, `attachMany`, `hasOne`, `hasMany`, `morphOne` and `morphMany`.
`->notDependent()` | Makes this relation independent of the primary model. The related model records will not be deleted when the primary model is deleted. This is the default behavior.
`->detachable()` | Makes this relation detach from the primary model if the primary model is deleted or the relationship is broken. This is the default behavior, and is only available, for the following relation types: `belongsToMany`, `morphToMany` and `morphedByMany`.
`->notDetachable()` | Prevents the relation from detaching from the primary model when the primary model is deleted or the relationship is broken.
`->pushable()` | Sets this relation to save when `push()` is run on the primary model. This is the default behavior.

> `->notPushable()` | Sets this relation to not be saved when `push()` is run on the primary model.

Reviewer: Thoughts / precedent for `x() / notX()` vs `x() / x(false)`?

Author: I find with chained methods that set flags like this, it's more readable to use just method names. This is my own view though, I don't have any specific precedent to cite for it. I added the `not...` methods purely to allow for people to reverse a flag in any sort of extension.

Reviewer: I think the pattern of having the toggle be in the parameters is fairly well established cross-language at this point, so I think my preference would be to not have the `notX()` methods present or documented. 

`->pushable(false)` vs `->notPushable()` seems to be about the same amount of readable to me. @mjauvin @jaxwilko any thoughts?

Reviewer 2: I'm good with `->pushable(true/false)`

Author: Can we have both? I do like how declarative method naming reads over having to use parameters and relying on the IDE to tell me what they mean - even if it's something as simple as an enable toggle. It's new API anyway, it's not gonna hurt 😛

Reviewer: @bennothommo sure, I'm fine with both

You might have noticed that there are no chain methods for handling "constraint" options like the `order`, `conditions` and `scope` options available in the property style relations. That is because relations defined in this format are already [query builders](query) - you can simply add the constraints directly to the relation!

```php
#[Relation]
public function posts()
{
    return $this->hasMany('Acme\Blog\Models\Post')->dependent()->published()->where('free_article', true);
}
```

> > **Note:** If your relation is constrained in this fashion, the object returned will be a query builder, not the original relation class. You will need to use the `Relation` attribute to mark methods that return query builders as a relation method.

Reviewer: ```suggestion
> **NOTE:** If your relation is constrained in this fashion, the object returned will be a query builder, not the original relation class. You will need to use the `Relation` attribute to mark methods that return query builders as a relation method.
```

> **Note:** Method style relation definition is available from Winter v1.2.5.

Model relations can also be defined as methods in a model class, similar to the base Laravel framework.

```php
class User extends Model
{
    public function profile(): HasOne
    {
        return $this->hasOne('Acme\Blog\Models\Profile');
    }

    public function posts(): HasMany
    {
        return $this->hasMany('Acme\Blog\Models\Post');
    }
}
```

This way provides a more familiar syntax for Laravel users, and allows for code editors and IDEs to provide syntax support for relations.

One key difference between property style definitions and method style definitions is that a relation method must *explicitly* be defined as a relation method in order for it to be picked up in certain circumstances, for example, when determining all available relations for cascading deletions.

In order to define a relation method, the method must have a single return type that matches one of the `Winter\Storm\Database\Relations` classes:

```php
use Winter\Storm\Database\Relations\HasMany;

public function posts(): HasMany
{
    return $this->hasMany('Acme\Blog\Models\Post');
}
```

> Alternatively, you may also use the `Relation` attribute on the method to define it is as a relation method:

Reviewer 2: ```suggestion
Alternatively, you may also use the `Relation` attribute on the method to define it as a relation method:
```

`->detachable()` | Makes this relation detach from the primary model if the primary model is deleted or the relationship is broken. This is the default behavior, and is only available, for the following relation types: `belongsToMany`, `morphToMany` and `morphedByMany`.
`->notDetachable()` | Prevents the relation from detaching from the primary model when the primary model is deleted or the relationship is broken.
`->pushable()` | Sets this relation to save when `push()` is run on the primary model. This is the default behavior.
`->notPushable()` | Sets this relation to not be saved when `push()` is run on the primary model.

You might have noticed that there are no chain methods for handling "constraint" options like the `order`, `conditions` and `scope` options available in the property style relations. That is because relations defined in this format are already [query builders](query) - you can simply add the constraints directly to the relation!

```php
public function posts(): HasMany
{
    return $this->hasMany('Acme\Blog\Models\Post')->dependent()->published()->where('free_article', true);

> }

Reviewer: Does this example actually work given that chaining the query methods will cause a builder to be returned instead of a relation instance?

Author: Good spot - no it probably wouldn't with the return type being `HasMany`, but it would if it were using the Relation attribute.

Author: Fixed that example.

class User extends Model
{
    // Property style
    public $hasOne = [
        'phone' => 'Acme\Blog\Models\Phone'
    ];

    // Method style
    public function phone(): HasOne
    {
        return $this->hasOne('Acme\Blog\Models\Phone');

>     }

Reviewer: What happens when someone double declares the relationship like this? Do we throw an exception warning them of this when parsing the property style config?

Author: The method function will supersede the property style config. Note that a lot of the functionality we have for property style relations is literally to emulate a relation method being available, so if the method exists, it will be used instead.

Reviewer: Can we make the property stuff throw an exception if it encounters this if it doesn't already? I wouldn't want someone to define a config in a property that isn't actually looked at because they have a method present.

Author: We wouldn't be able to catch all cases of this. If, say, there's a `phone` definition in the relation properties, and a `phone` method in the same model, and you were to call `$model->phone()`, the method is going to be called without issue because it exists.

It would only be in cases where you were to call `$model->phone` (to ostensibly get the "value" of the relation) that we could possibly pick it up, because as you can see, the other way of calling the relation is naturally going to go to the method over the property.

Reviewer 2: @bennothommo what about calling it in twig without "()" ? Does it default to the method or the property ?

Reviewer: Let's add the logic to detect it anyways, better to have some protection in place than none

Author: @mjauvin it would default to the property first in Twig based on its access logic rules.

Author: Done as of https://github.com/wintercms/storm/pull/167/commits/55603fc00f5780bd7805e42bce4a209305ef9c02

    }
}
```

In the example above, the model will try to match the `user_id` from the `Phone` model to an `id` on the `User` model. It determines the default foreign key name by examining the name of the relationship definition and suffixing the name with `_id`. However, if the foreign key on the `Phone` model is not `user_id`, you may pass a custom key name using the `key` parameter on the definition:

```php
// Property style
public $belongsTo = [
    'user' => ['Acme\Blog\Models\User', 'key' => 'my_user_id']
];

// Method style
public function user(): BelongsTo
{
    return $this->belongsTo('Acme\Blog\Models\User', 'my_user_id');
}
```

If your parent model does not use `id` as its primary key, or you wish to join the child model to a different column, you may pass the `otherKey` parameter to the definition specifying your parent table's custom key:

```php
// Property style
public $belongsTo = [
    'user' => ['Acme\Blog\Models\User', 'key' => 'my_user_id', 'otherKey' => 'my_id']
];

// Method style
public function user(): BelongsTo
{
    return $this->belongsTo('Acme\Blog\Models\User', 'my_user_id', 'my_id');
}
```

#### Default models

> The `belongsTo`, `hasOne`, `hasOneThrough` and `morphOne` relationships allow you to define a default model that will be returned if the given relationship is `null`. This pattern is often referred to as the [Null Object pattern](https://en.wikipedia.org/wiki/Null_Object_pattern) and can help remove conditional checks in your code. In the following example, the `user` relation will return an empty `Acme\Blog\Models\User` model if no `user` is attached to the post:

Reviewer: ```suggestion
The `belongsTo`, `hasOne`, `hasOneThrough`, and `morphOne` relationships allow you to define a default model that will be returned if the given relationship is `null`. This pattern is often referred to as the [Null Object pattern](https://en.wikipedia.org/wiki/Null_Object_pattern) and can help remove conditional checks in your code. In the following example, the `user` relation will return an empty `Acme\Blog\Models\User` model if no `user` is attached to the post:
```

$comment->save();
```

## Dynamically defining a relation

Using Winter's powerful [extension capabilities](../services/behaviors), you can dynamically add additional relations to models at run-time, in both the property style and the method style.

Within a [plugin boot method](../plugin/registration#registration-file), you can extend a given model to add additional relations:

```php
<?php

namespace Acme\Blog;

use Winter\Storm\Database\Relations\HasMany;

class Plugin extends \System\Classes\PluginBase
{
    // ...
    public function boot()
    {
        \Acme\Blog\Post::extend(function ($model) {
            // Property-style
            $model->hasMany['comments'] = [
                \Acme\Blog\Comment::class,
            ];

            // Method-style
            $model->addDynamicMethod('comments', function (): HasMany {
                return $model->hasMany(\Acme\Blog\Comment::class);
            });
        });
    }

> }

Reviewer: Last thing required, need to document the methods I added a couple years ago to make adding new relationships easier: https://github.com/wintercms/storm/commit/b977610d361cc25f24b05062e40a83a6d53680a8
